test/e2e/scenario/recorder/viewports.scenario.ts

```diff
@@ -0,0 +1,371 @@
+      actions: [
+        { type: 'pointerMove', duration: 0, x: xBase + xOffsetFingerTwo, y: yBase },
+        { type: 'pointerDown', button: 0 },
+        { type: 'pause', duration: pauseDurationMs },
+        { type: 'pointerMove', duration: moveDurationMs, origin: 'pointer', x: +xChange, y: 0 },
+        { type: 'pointerUp', button: 0 },
+      ],
+    },
+  ]
+  await driver.performActions(actions)
+}
+
+async function performSignificantZoom() {
+  const initialVisualViewport = await getVisualViewport()
+  await pinchZoom(150)
+  await pinchZoom(150)
+  const nextVisualViewport = await getVisualViewport()
+  // Test the test: ensure pinch zoom was applied
+  expect(initialVisualViewport.scale < nextVisualViewport.scale).toBeTruthy()
+}
+
+async function pinchScrollVerticallyDown(yChange: number) {
+  // Providing a negative offset value will scroll up.
+  // NOTE: Some devices may invert scroll direction
+  // Cannot exceed the bounds of a device's screen, at start or end positions.
+  // So pick a midpoint on small devices, roughly 180px.
+  const xBase = 180
+  const yBase = 180
+  // Scrolling too fast can show or hide the address bar on some device browsers.
+  const moveDurationMs = 800
+  const pauseDurationMs = 150
+
+  const actions = [
+    {
+      type: 'pointer',
+      id: 'finger1',
+      parameters: { pointerType: 'touch' },
+      actions: [
+        { type: 'pointerMove', duration: 0, x: xBase, y: yBase },
+        { type: 'pointerDown', button: 0 },
+        { type: 'pause', duration: pauseDurationMs },
+        { type: 'pointerMove', duration: moveDurationMs, origin: 'pointer', x: 0, y: -yChange },
+        { type: 'pointerUp', button: 0 },
+      ],
+    },
+  ]
+  await driver.performActions(actions)
+}
+
+async function buildScrollablePage() {
+  await browserExecute(() => {
+    document.documentElement.style.setProperty('width', '5000px')
+    document.documentElement.style.setProperty('height', '5000px')
+    document.documentElement.style.setProperty('margin', '0px')
+    document.documentElement.style.setProperty('padding', '0px')
+    document.body.style.setProperty('margin', '0px')
+    document.body.style.setProperty('padding', '0px')
+    document.body.style.setProperty('width', '5000px')
+    document.body.style.setProperty('height', '5000px')
+  })
+}
+
+interface VisualViewportData {
+  scale: number
+  width: number
+  height: number
+  offsetLeft: number
+  offsetTop: number
+  pageLeft: number
+  pageTop: number
+}
+
+function getVisualViewport(): Promise<VisualViewportData> {
+  return browserExecute(() => {
+    const visual = window.visualViewport || {}
+    return {
+      scale: visual.scale,
+      width: visual.width,
+      height: visual.height,
+      offsetLeft: visual.offsetLeft,
+      offsetTop: visual.offsetTop,
+      pageLeft: visual.pageLeft,
+      pageTop: visual.pageTop,
+    }
+  }) as Promise<VisualViewportData>
+}
+
+function getWindowScroll() {
```

Comment:
```suggestion
function getLayoutScroll() {
```
to be consistent with `resetLayoutScroll` naming

Reply:
Unfortunately browsers are not consistent on the `window.scrollX/Y` specification. This could be either layout scroll, or layout+visual scroll. For the devices we test on, they happen to be consistent- but I'd recommend we stick with `getWindowScroll` to make it clear that it is the native behaviour- which is loosely defined.

Reply:
so `resetWindowScroll` then?

Reply:
Funny enough `resetLayoutScroll` does in fact reset the layout scroll though has a somewhat poorly defined affect on visual scroll. But yes, this would follow.